spring-graphql/src/main/java/org/springframework/graphql/data/method/annotation/support/AuthenticationPrincipalArgumentResolver.java

```diff
@@ -0,0 +1,145 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.graphql.data.method.annotation.support;
+
+import graphql.schema.DataFetchingEnvironment;
+import org.reactivestreams.Publisher;
+import org.springframework.core.MethodParameter;
+import org.springframework.core.ResolvableType;
+import org.springframework.core.annotation.AnnotationUtils;
+import org.springframework.expression.BeanResolver;
+import org.springframework.expression.Expression;
+import org.springframework.expression.ExpressionParser;
+import org.springframework.expression.spel.standard.SpelExpressionParser;
+import org.springframework.expression.spel.support.StandardEvaluationContext;
+import org.springframework.graphql.data.method.HandlerMethodArgumentResolver;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.annotation.AuthenticationPrincipal;
+import org.springframework.security.core.context.ReactiveSecurityContextHolder;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.StringUtils;
+import reactor.core.publisher.Mono;
+
+import java.lang.annotation.Annotation;
+
+/**
+ * Resolver to obtain {@link Authentication#getPrincipal()} from Spring Security context via
+ * {@link SecurityContext#getAuthentication()} for parameters annotated with {@link AuthenticationPrincipal}.
+ *
+ * <p>The resolver checks both ThreadLocal context via {@link SecurityContextHolder}
+ * for Spring MVC applications, and {@link ReactiveSecurityContextHolder} for
+ * Spring WebFlux applications.
+ *
+ * @author Rob Winch
+ * @since 1.0.0
+ */
+public class AuthenticationPrincipalArgumentResolver implements HandlerMethodArgumentResolver {
+
+	private ExpressionParser parser = new SpelExpressionParser();
+
+	private BeanResolver beanResolver;
+
+	/**
+	 * Sets the {@link BeanResolver} to be used on the expressions
+	 * @param beanResolver the {@link BeanResolver} to use
+	 */
+	public void setBeanResolver(BeanResolver beanResolver) {
+		this.beanResolver = beanResolver;
+	}
+
+	@Override
+	public boolean supportsParameter(MethodParameter parameter) {
+		return findMethodAnnotation(AuthenticationPrincipal.class, parameter) != null;
+	}
+
+	@Override
+	public Object resolveArgument(MethodParameter parameter, DataFetchingEnvironment environment) throws Exception {
+		return getCurrentAuthentication().map(Authentication::getPrincipal)
+				.map((principal) -> resolvePrincipal(parameter, principal))
+				.transform((argument) -> {
+					Class<?> parameterType = parameter.getParameterType();
+					boolean isParameterPublisher = Publisher.class.isAssignableFrom(parameterType);
```

Comment:
Something like this would be a  more accurate check:
```java
private boolean isMonoAssignable(MethodParameter parameter) {
	return (parameter.getParameterType().equals(Publisher.class) ||
			parameter.getParameterType().equals(Mono.class));
}
```
In other words it has to be Publisher or Mono, and any other Publisher type it would be a ClassCastException. Unlikely indeed, I think a check for equals `Mono` would be sufficient.

Reply:
This is now done